contracts/services/reverse-splitter/src/contract.rs

```diff
@@ -0,0 +1,299 @@
+        querier: &QuerierWrapper<Empty>,
+    ) -> Result<
+        Vec<(
+            cosmwasm_std::Uint128,
+            &'a valence_service_utils::denoms::CheckedDenom,
+            &'a Addr,
+        )>,
+        ServiceError,
+    > {
+        // Get input account balances for each denom (one balance query per denom)
+        let mut account_balances: HashMap<String, Uint128> = HashMap::new();
+        // Dynamic ratios
+        let mut dynamic_ratios: HashMap<String, Decimal> = HashMap::new();
+        let mut denom_amount_count = 0;
+
+        for split in cfg.splits() {
+            let denom = split.denom();
+            let key = account_key(split.account(), denom);
+            // Query account/denom balance and add it to cache
+            let balance = denom.query_balance(querier, split.account())?;
+            if let Some(amount) = split.amount() {
+                // Stop if the specified amount is greater than the input account's balance
+                if *amount > balance {
+                    return Err(ServiceError::ExecutionError(format!(
+                        "Insufficient balance on account {} for denom '{:?}' in split config (required: {}, available: {}).",
+                        split.account(), denom, amount, balance,
+                    )));
+                }
+                denom_amount_count += 1;
+            }
+            account_balances.insert(key, balance);
+
+            if let Some(RatioConfig::DynamicRatio {
+                contract_addr,
+                params,
+            }) = split.ratio()
+            {
+                let key = dyn_ratio_key(denom, contract_addr, params);
+                if let Entry::Vacant(e) = dynamic_ratios.entry(key) {
+                    let ratio = query_dynamic_ratio(querier, contract_addr, params, denom)?;
+                    e.insert(ratio);
+                }
+            }
+        }
+
+        if denom_amount_count == cfg.splits().len() {
+            // If all splits have an amount (and we have checked the balances),
+            // we can return the amounts as-is.
+            return cfg
+                .splits()
+                .iter()
+                .map(|split| Ok((split.amount().unwrap(), split.denom(), split.account())))
+                .collect::<Result<Vec<_>, _>>();
+        }
+
+        // If not all splits have an amount, we need to compute the amounts based on the ratios
+
+        // Prepare transfer messages for each split config
+        let amounts_in_base_denom = cfg
+            .splits()
+            .iter()
+            .map(|split| {
+                // Lookup account/denom balance
+                let account = split.account();
+                let denom = split.denom();
+                let balance = account_balances.get(&account_key(account, denom)).unwrap();
+                split
+                    .amount()
+                    .map(|amount| {
+                        assert_eq!(denom, cfg.base_denom());
```

Comment:
remove this? It has been already validated no? Or if you want to double check return a ServiceError instead?

Reply:
Yep .. I'll remove it.